File: src/brs/web/api/http/handler/IssueAsset.java
package brs.web.api.http.handler;

import brs.*;
import brs.fluxcapacitor.FluxValues;
import brs.services.ParameterService;
import brs.util.Convert;
import brs.util.TextUtils;
import brs.web.api.http.common.APITransactionManager;
import brs.web.api.http.common.JSONResponses;
import brs.web.api.http.common.LegacyDocTag;
import brs.web.api.http.common.ParameterParser;
import com.google.gson.JsonElement;

import javax.servlet.http.HttpServletRequest;

import static brs.web.api.http.common.JSONResponses.*;
import static brs.web.api.http.common.Parameters.*;

public final class IssueAsset extends CreateTransaction {

  private final ParameterService parameterService;
  private final Blockchain blockchain;

  public IssueAsset(ParameterService parameterService, Blockchain blockchain, APITransactionManager apiTransactionManager) {
    super(new LegacyDocTag[]{LegacyDocTag.AE, LegacyDocTag.CREATE_TRANSACTION}, apiTransactionManager,
        NAME_PARAMETER, DESCRIPTION_PARAMETER, QUANTITY_QNT_PARAMETER, DECIMALS_PARAMETER, MINTABLE_PARAMETER);
    this.parameterService = parameterService;
    this.blockchain = blockchain;
  }

  @Override
  protected
  JsonElement processRequest(HttpServletRequest req) throws BurstException {

    String name = req.getParameter(NAME_PARAMETER);
    String description = req.getParameter(DESCRIPTION_PARAMETER);
    String decimalsValue = Convert.emptyToNull(req.getParameter(DECIMALS_PARAMETER));
    boolean mintable = "true".equals(req.getParameter(MINTABLE_PARAMETER));
    if(mintable && !Signum.getFluxCapacitor().getValue(FluxValues.SMART_TOKEN)) {
      //only after the fork we are allowed to have a mintable assset
      return JSONResponses.incorrect(MINTABLE_PARAMETER);
    }

    if (name == null) {
      return MISSING_NAME;
    }

    name = name.trim();
    if (name.length() < Constants.MIN_ASSET_NAME_LENGTH || name.length() > Constants.MAX_ASSET_NAME_LENGTH) {
      return INCORRECT_ASSET_NAME_LENGTH;
    }

    if (!TextUtils.isInAlphabet(name)) {
      return INCORRECT_ASSET_NAME;
    }

    if (description != null && description.length() > Constants.MAX_ASSET_DESCRIPTION_LENGTH) {
      return INCORRECT_ASSET_DESCRIPTION;
    }

    byte decimals = 0;
    if (decimalsValue != null) {
      try {
        decimals = Byte.parseByte(decimalsValue);
        if (decimals < 0 || decimals > 8) {
          return INCORRECT_DECIMALS;
        }
      } catch (NumberFormatException e) {
        return INCORRECT_DECIMALS;
      }
    }

    long quantityQNT = ParameterParser.getQuantityQNT(req);
    if(quantityQNT == 0L && !mintable) {
      return incorrect(QUANTITY_QNT_PARAMETER);
    }
    Account account = parameterService.getSenderAccount(req);
    Attachment attachment = new Attachment.ColoredCoinsAssetIssuance(name, description, quantityQNT, decimals, blockchain.getHeight(), mintable);
    return createTransaction(req, account, attachment);

  }

}

File: src/main/kotlin/brs/http/IssueAsset.kt
package brs.http

import brs.Attachment
import brs.Constants
import brs.DependencyProvider
import brs.http.JSONResponses.INCORRECT_ASSET_DESCRIPTION
import brs.http.JSONResponses.INCORRECT_ASSET_NAME
import brs.http.JSONResponses.INCORRECT_ASSET_NAME_LENGTH
import brs.http.JSONResponses.INCORRECT_DECIMALS
import brs.http.JSONResponses.MISSING_NAME
import brs.http.common.Parameters.DECIMALS_PARAMETER
import brs.http.common.Parameters.DESCRIPTION_PARAMETER
import brs.http.common.Parameters.NAME_PARAMETER
import brs.http.common.Parameters.QUANTITY_QNT_PARAMETER
import brs.util.TextUtils
import brs.util.convert.emptyToNull
import com.google.gson.JsonElement
import javax.servlet.http.HttpServletRequest

internal class IssueAsset internal constructor(private val dp: DependencyProvider) : CreateTransaction(dp, arrayOf(APITag.AE, APITag.CREATE_TRANSACTION), NAME_PARAMETER, DESCRIPTION_PARAMETER, QUANTITY_QNT_PARAMETER, DECIMALS_PARAMETER) {

    override fun processRequest(request: HttpServletRequest): JsonElement {

        var name: String? = request.getParameter(NAME_PARAMETER)
        val description = request.getParameter(DESCRIPTION_PARAMETER)
        val decimalsValue = request.getParameter(DECIMALS_PARAMETER).emptyToNull()

        if (name == null) {
            return MISSING_NAME
        }

        name = name.trim { it <= ' ' }
        if (name.length < Constants.MIN_ASSET_NAME_LENGTH || name.length > Constants.MAX_ASSET_NAME_LENGTH) {
            return INCORRECT_ASSET_NAME_LENGTH
        }

        if (!TextUtils.isInAlphabet(name)) {
            return INCORRECT_ASSET_NAME
        }

        if (description != null && description.length > Constants.MAX_ASSET_DESCRIPTION_LENGTH) {
            return INCORRECT_ASSET_DESCRIPTION
        }

        var decimals: Byte = 0
        if (decimalsValue != null) {
            try {
                decimals = java.lang.Byte.parseByte(decimalsValue)
                if (decimals < 0 || decimals > 8) {
                    return INCORRECT_DECIMALS
                }
            } catch (e: NumberFormatException) {
                return INCORRECT_DECIMALS
            }

        }

        val quantityQNT = ParameterParser.getQuantityQNT(request)
        val account = dp.parameterService.getSenderAccount(request)
        val attachment = Attachment.ColoredCoinsAssetIssuance(dp, name, description!!, quantityQNT, decimals, dp.blockchain.height)
        return createTransaction(request, account, attachment)
    }
}
